File: Code/SlidingWindow/Minimum_Swaps_to_Group_All_1_Together_II_2134.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution
{
public:
    /*
        the point according to question we want all ones together right,
        so first we calculate total ones, with this totalones window we calculte max one present in windowOfOnes.size() .
        Because the end have one count to be in same group we make array circular, how?
        sizeofArray*2 and access element in i%sizeofArray
        and finally minus all the maxOnes in windowOfOnes.size() from totalOnes

    */
    int minSwaps(vector<int> nums)
    {
        int ttlOnes = count(nums.begin(), nums.end(), 1);
        int n = nums.size();
        int groupOfOnes = 0, maxGroupOfOnes = 0;
        for (int i = 0, j = 0; i < n * 2; i++)
        {
            if (nums[i % n])
                ++groupOfOnes;

            if (i >= ttlOnes)
            {
                if (nums[j % n])
                    --groupOfOnes;
                maxGroupOfOnes = max(maxGroupOfOnes, groupOfOnes);
                ++j;
            }
        }

        return ttlOnes - maxGroupOfOnes;
    }
};
int main()
{
#ifndef ONLINE_JUDGE
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);
#endif
    Solution obj;
    //

    auto t = obj.minSwaps({0, 1, 1, 1, 0, 0, 1, 1, 0});
    cout << t;
    return 0;
}

File: java/SlidingWindow_java/Minimum_Swaps_to_Group_All_1_Together_II_2134.java
package java.SlidingWindow_java;

// import static java.lang.Math.max;
// import static java.lang.Math.min;
// import static java.lang.Math.abs;
// import static java.lang.System.out;
import java.util.*;
import java.io.*;

public class Minimum_Swaps_to_Group_All_1_Together_II_2134 {
    /*
     * the point according to question we want all ones together right,
     * so first we calculate total ones, with this totalones window we calculte max
     * one present in windowOfOnes.size() .
     * Because the end have one count to be in same group we make array
     * circular,how?
     * sizeofArray*2 and access element in i%sizeofArray
     * and finally minus all the maxOnes in windowOfOnes.size() from totalOnes
     * 
     */
    public int minSwaps(int[] nums) {
        int ttlOnes = Arrays.stream(nums).sum();
        int n = nums.length;
        int groupOfOnes = 0, maxGroupOfOnes = 0;
        for (int i = 0, j = 0; i < n * 2; i++) {
            if (nums[i % n] == 1)
                ++groupOfOnes;

            if (i >= ttlOnes) {
                if (nums[j % n] == 1)
                    --groupOfOnes;
                maxGroupOfOnes = Math.max(maxGroupOfOnes, groupOfOnes);
                ++j;
            }
        }

        return ttlOnes - maxGroupOfOnes;
    }

    public static void main(String[] args) throws FileNotFoundException {
        File f = new File("output.txt");

        // Scanner ip = new Scanner(System.in);
        try (Scanner ip = new Scanner(new File("input.txt"));
                PrintWriter op = new PrintWriter(f);) {

            // int x = ip.nextInt();
            // int y = ip.nextInt();
            // op.println(x + y);

            int grid[] = { 1, 1, 1, 0, 0, 1, 0, 1, 1, 0 };
            Minimum_Swaps_to_Group_All_1_Together_II_2134 ob = new Minimum_Swaps_to_Group_All_1_Together_II_2134();
            int ans = ob.minSwaps(grid);
            op.println(ans);
        }
    }
}